miraclelinux/meta-debian-extended
# Add kexec-tools package.

# Purpose of pull request

kexec-tools 

This pr adds kexec-toolspackage based on poky's meta/recipes-kernel/kexec/kexec-tools_2.0.19.bb.

# Test
## How to test

ref. https://wiki.archlinux.jp/index.php/Kexec
https://wiki.archlinux.jp/index.php/Kdump

Test on qemu64.
Build an image for testing.
```
$ echo 'IMAGE_INSTALL_append = " kexec kdump vmcore-dmesg "' >> conf/local.conf
$ echo 'IMAGE_INSTALL_append = " openssh "' >> conf/local.conf
$ echo 'IMAGE_ROOTFS_EXTRA_SPACE_append = " + 1000000"' >> conf/local.conf
```
After bitabake is finished, start QEMU with the following command.
```
$ MACHINE=qemuarm64 runqemu nographic bootparams="crashkernel=128M"
```
After starting qemu, launch a new terminal in the host environment and transfer the Image from the host environment.
- For EMLINUX, specify the path to emlinux
- The IP address of qemu should match the environment.
```
$ scp ${EMLINUX}/build/tmp-glibc/deploy/images/qemuarm64/Image root@192.168.7.2:/boot/
```

### kexec
Using the kexec command, load the kernel image and confirm the boot.
```
root@qemuarm64:/boot# kexec -l /boot/Image --reuse-cmdline
root@qemuarm64:/boot# kexec -e
```
### kdump
Partially edited the kdump init script /etc/init.d/kdump for the test environment.
The differences before and after the edit are as follows:
```
root@qemuarm64:/etc/init.d# diff kdump.orig kdump
--- kdump.orig
+++ kdump
@@ -15,6 +15,8 @@
 MAKEDUMPFILE_ARGS="-E -d 1"

 LOGGER="logger -p info -t kdump"
+KDUMP_KIMAGE="/boot/Image"
+KDUMP_CMDLINE="`cat /proc/cmdline`"

 if [ -f /etc/sysconfig/kdump.conf ]; then
        . /etc/sysconfig/kdump.conf
@@ -26,10 +28,10 @@
 do_check()
 {
        #check makedumpfile
-       if [ ! -e ${MAKEDUMPFILE} -o ! -x ${MAKEDUMPFILE} ] ;then
-               echo "No makedumpfile found."
-               exit 0
-       fi
+#      if [ ! -e ${MAKEDUMPFILE} -o ! -x ${MAKEDUMPFILE} ] ;then
+#              echo "No makedumpfile found."
+#              exit 0
+#      fi

        #check kexec
        if [ ! -e ${KEXEC} -o ! -x ${KEXEC} ] ;then
@@ -58,8 +60,9 @@
        mkdir -p ${KDUMP_VMCORE_PATH}
        echo "Saving a vmcore to ${KDUMP_VMCORE_PATH}."

-       ${MAKEDUMPFILE} ${MAKEDUMPFILE_ARGS} /proc/vmcore ${KDUMP_VMCORE_PATH}/vmcore-"`date +"%H:%M:%S"`"
+#      ${MAKEDUMPFILE} ${MAKEDUMPFILE_ARGS} /proc/vmcore ${KDUMP_VMCORE_PATH}/vmcore-"`date +"%H:%M:%S"`"
 #      cp --sparse=always /proc/vmcore ${KDUMP_VMCORE_PATH}/vmcore-"`date +"%H:%M:%S"`"
+       cp /proc/vmcore ${KDUMP_VMCORE_PATH}/vmcore-"`date +"%H:%M:%S"`"
        rc=$?
        if [ ${rc} == 0 ]; then
                ${LOGGER} "Saved a vmcore to ${KDUMP_VMCORE_PATH}."
```
Reboot after editing
After rebooting, make sure that the kernel for the crash is already loaded, and then trigger the crash.
```
root@qemuarm64:~# cat /sys/kernel/kexec_crash_loaded
1                      # <- "1" is loaded.
root@qemuarm64:~# echo c > /proc/sysrq-trigger
### Boot with the kexec kernel, save the dump file, and then reboot again.
```
After rebooting, log in and check that the dump file has been saved.

## Test result

### kexec
```
root@qemuarm64:/boot# kexec -l /boot/Image --reuse-cmdline
root@qemuarm64:/boot# kexec -e
[  561.725066] kexec_core: Starting new kernel
[  561.726143] Disabling non-boot CPUs ...
[  561.740694] Bye!
[    0.000000] Booting Linux on physical CPU 0x0000000000 [0x411fd070]
[    0.000000] Linux version 4.19.205-cip56 (dev@abcd-efg-hijklmn) (gcc version 8.3.0 (GCC)) #1 SMP PREEMPT Wed Nov 24 15:01:34 JST 2021
[    0.000000] Machine model: linux,dummy-virt
[    0.000000] Memory limited to 512MB
... <snip> ...
EMLinux 2.4 qemuarm64 /dev/ttyAMA0

qemuarm64 login:
```
### kdump
```
root@qemuarm64:~# cat /sys/kernel/kexec_crash_loaded
1
root@qemuarm64:~# echo c > /proc/sysrq-trigger
[  133.497862] sysrq: Trigger a crash
[  133.503289] Unable to handle kernel NULL pointer dereference at virtual address 0000000000000000
[  133.503610] Mem abort info:
[  133.503734]   ESR = 0x96000046
...
```
```
root@qemuarm64:~# ls -lh /var/crash/2021-11-24/
-r--------    1 root     root      403.6M Nov 24 07:56 vmcore-07:55:54
```
